src/swarm-tui/Swarm/TUI/View/Robot.hs

```diff
@@ -28,17 +42,223 @@ import Swarm.Game.State.Substate
-import Swarm.TUI.View.Util as VU
-import Swarm.Util
+import Swarm.TUI.View.Robot.Type
+import Swarm.Util (applyWhen, maximum0, maximumNE)
 import Swarm.Util.UnitInterval
 import Swarm.Util.WindowedCounter qualified as WC
 import System.Clock (TimeSpec (..))
 
+extractColWidth :: ColWidth -> Int
+extractColWidth (ColW x) = x
+
+instance Semigroup ColWidth where
+  ColW w1 <> ColW w2 = ColW $ max w1 w2
+
+data RobotRenderingContext = RobotRenderingContext
+  { _mygs :: GameState
+  , _gameplay :: UIGameplay
+  , _timing :: UITiming
+  , _uiDbg :: Set DebugOption
+  }
+
+makeLenses ''RobotRenderingContext
+
+mkRobotDisplay :: RobotRenderingContext -> MixedTabularList Name RobotWidgetRow Widths
+mkRobotDisplay c =
+  mixedTabularList
+    (RobotsListDialog RobotList)
+    (mkLibraryEntries c)
+    (LstItmH 1)
+    (computeColumnWidths $ c ^. uiDbg)
+    assignRowWidthForConstructor
+
+emptyRobotDisplay :: Set DebugOption -> RobotListContent
+emptyRobotDisplay uiDebug =
+  RobotListContent
+    { _robotsListWidget = mixedTabularList (RobotsListDialog RobotList) mempty (LstItmH 1) (computeColumnWidths uiDebug) assignRowWidthForConstructor
+    , _robotsListRenderers =
+        MixedRenderers
+          { cell = drawCell uiDebug
+          , rowHdr = Just rowHdr
+          , colHdr = Just $ colHdr uiDebug
+          , colHdrRowHdr = Just $ ColHdrRowHdr $ \_ _ -> vLimit 1 (fill ' ') <=> hBorder
+          }
+    , _robotDetailsPaneState =
+        RobotDetailsPaneState
+          { _logsList = BL.list (RobotsListDialog $ SingleRobotDetails RobotLogPane) mempty 1
+          , _cmdHistogramList = BL.list (RobotsListDialog $ SingleRobotDetails RobotCommandHistogramPane) mempty 1
+          }
+    }
+
+renderRobotsList :: RobotListContent -> Widget Name
+renderRobotsList rd =
+  vLimit 30 $
+    renderMixedTabularList (rd ^. robotsListRenderers) (LstFcs True) (rd ^. robotsListWidget)
+
+columnHdrAttr :: AttrName
+columnHdrAttr = attrName "columnHeader"
+
+rowHdrAttr :: AttrName
+rowHdrAttr = attrName "rowHeader"
+
+colHdr :: Set DebugOption -> MixedColHdr Name Widths
+colHdr uiDebug =
+  MixedColHdr
+    { draw = \_ (MColC (Ix ci)) -> case hdrs V.!? ci of
+        Just ch -> withAttr columnHdrAttr (str ch) <=> hBorder
+        Nothing -> emptyWidget
+    , widths = \(Widths ws) -> zipWith (<>) ws (map (ColW . length) $ V.toList hdrs)
+    , height = ColHdrH 2
+    }
+ where
+  hdrs = colHdrs uiDebug
+
+-- | Enumerates the rows by position (not 'RID').
+rowHdr :: RowHdr Name RobotWidgetRow
+rowHdr =
+  RowHdr
+    { draw = \_ (WdthD wd) (RowHdrCtxt (Sel s)) rh ->
+        let attrFn =
+              if s
+                then id
+                else withAttr rowHdrAttr
+         in attrFn $ padRight (Pad $ if wd > 0 then 0 else 1) $ padLeft Max (str $ show rh)
+    , width = \_ rh -> RowHdrW . (+ 2) . maximum0 $ map (length . show) rh
+    , toRH = \_ (Ix i) -> i + 1
+    }
+
+data ColumnExpansion
+  = Grow
+  | Minimal
+  deriving (Eq)
```

Comment:
Why not use [`Brick.Types.Size`](https://hackage.haskell.org/package/brick-2.4/docs/Brick-Types.html#t:Size)?

Reply:
Added a docstring.